maven-core/src/main/java/org/apache/maven/project/ReactorModelPool.java

```diff
@@ -19,53 +19,120 @@
+     * @throws NoSuchElementException if model could not be found
+     */
+    public Model get( String groupId, String artifactId, String version )
+        throws IllegalStateException, NoSuchElementException
+    {
+        return modelsByGa.getOrDefault( new GAKey( groupId, artifactId ), Collections.emptySet() ).stream()
+                        .filter( m -> version == null || version.equals( getVersion( m ) ) )
+                        .reduce( ( a, b ) -> 
+                        {
+                            throw new IllegalStateException( "Multiple modules with key "
+                                + a.getGroupId() + ':' + a.getArtifactId() );
+                        } ).orElse( null );
+    }
 
-    private final Map<CacheKey, File> pomFiles = new HashMap<>();
-
-    public File get( String groupId, String artifactId, String version )
+    public Model get( Path path )
     {
-        return pomFiles.get( new CacheKey( groupId, artifactId, version ) );
+        final Path pomFile;
+        if ( Files.isDirectory( path ) )
+        {
+            pomFile = path.resolve( "pom.xml" );
+        }
+        else
+        {
+            pomFile = path;
+        }
+        return modelsByPath.get( pomFile );
+    }
+    
+    private String getVersion( Model model )
+    {
+        String version = model.getVersion();
+        if ( version == null && model.getParent() != null )
+        {
+            version = model.getParent().getVersion();
+        }
+        return version;
     }
 
-    public void put( String groupId, String artifactId, String version, File pomFile )
+    static class Builder
     {
-        pomFiles.put( new CacheKey( groupId, artifactId, version ), pomFile );
+        private ReactorModelPool pool = new ReactorModelPool();
+        
+        Builder put( Path pomFile, Model model )
+        {
+            pool.modelsByPath.put( pomFile, model );
+            pool.modelsByGa.computeIfAbsent( new GAKey( getGroupId( model ), model.getArtifactId() ),
+                                             k -> new HashSet<Model>() ).add( model );
+            return this;
+        }
+        
+        ReactorModelPool build() 
+        {
+            return pool;
+        }
+
+        private static String getGroupId( Model model )
+        {
+            String groupId = model.getGroupId();
+            if ( groupId == null && model.getParent() != null )
+            {
+                groupId = model.getParent().getGroupId();
+            }
+            return groupId;
+        }
     }
 
-    private static final class CacheKey
+    private static final class GAKey
     {
 
         private final String groupId;
 
         private final String artifactId;
 
-        private final String version;
-
         private final int hashCode;
 
-        CacheKey( String groupId, String artifactId, String version )
+        GAKey( String groupId, String artifactId )
```

Comment:
Why did you remove the version from here?

Reply:
It will be stored without version into `private final Map<GAKey, Set<Model>>`, so you can still get it including the version. This makes it possible to get the dependency (without version) from the reactor if and only if there's one version.  During build Maven ensures you're picking up reactor files, while the uploaded pom will inject this version.